test/units/module_utils/basic/test_known_hosts.py

```diff
@@ -19,37 +19,88 @@
+             'add_host_key_cmd': " -t rsa four.example.org",
+             'port': None},
         'git+ssh://five.example.org/example.git':
-            {'is_ssh_url': True, 'get_fqdn': 'five.example.org'},
-        'ssh://six.example.org:21/example.org':
-            {'is_ssh_url': True, 'get_fqdn': 'six.example.org'},
+            {'is_ssh_url': True, 'get_fqdn': 'five.example.org',
+             'add_host_key_cmd': " -t rsa five.example.org",
+             'port': None},
+        'ssh://six.example.org:21/example.org':  # ssh on FTP Port?
+            {'is_ssh_url': True, 'get_fqdn': 'six.example.org',
+             'add_host_key_cmd': " -t rsa -p 21 six.example.org",
+             'port': '21'},
         'ssh://[2001:DB8::abcd:abcd]/example.git':
-            {'is_ssh_url': True, 'get_fqdn': '[2001:DB8::abcd:abcd]'},
+            {'is_ssh_url': True, 'get_fqdn': '[2001:DB8::abcd:abcd]',
+             'add_host_key_cmd': " -t rsa [2001:DB8::abcd:abcd]",
+             'port': None},
         'ssh://[2001:DB8::abcd:abcd]:22/example.git':
-            {'is_ssh_url': True, 'get_fqdn': '[2001:DB8::abcd:abcd]'},
+            {'is_ssh_url': True, 'get_fqdn': '[2001:DB8::abcd:abcd]',
+             'add_host_key_cmd': " -t rsa -p 22 [2001:DB8::abcd:abcd]",
+             'port': '22'},
         'username@[2001:DB8::abcd:abcd]/example.git':
-            {'is_ssh_url': True, 'get_fqdn': '[2001:DB8::abcd:abcd]'},
-        'username@[2001:DB8::abcd:abcd]:22/example.git':
-            {'is_ssh_url': True, 'get_fqdn': '[2001:DB8::abcd:abcd]'},
+            {'is_ssh_url': True, 'get_fqdn': '[2001:DB8::abcd:abcd]',
+             'add_host_key_cmd': " -t rsa [2001:DB8::abcd:abcd]",
+             'port': None},
+        'username@[2001:DB8::abcd:abcd]:path/example.git':
+            {'is_ssh_url': True, 'get_fqdn': '[2001:DB8::abcd:abcd]',
+             'add_host_key_cmd': " -t rsa [2001:DB8::abcd:abcd]",
+             'port': None},
+        'ssh://internal.git.server:7999/repos/repo.git':
+            {'is_ssh_url': True, 'get_fqdn': 'internal.git.server',
+             'add_host_key_cmd': " -t rsa -p 7999 internal.git.server",
+             'port': '7999'}
     }
 
     def test_is_ssh_url(self):
         for u in self.urls:
             self.assertEqual(known_hosts.is_ssh_url(u), self.urls[u]['is_ssh_url'])
 
-    def test_get_fqdn(self):
+    def test_get_fqdn_and_port(self):
         for u in self.urls:
-            self.assertEqual(known_hosts.get_fqdn(u), self.urls[u]['get_fqdn'])
+            self.assertEqual(known_hosts.get_fqdn_and_port(u), (self.urls[u]['get_fqdn'], self.urls[u]['port']))
+
+    def test_add_host_key(self):
 
+        # Copied
+        args = json.dumps(dict(ANSIBLE_MODULE_ARGS={}))
+        # unittest doesn't have a clean place to use a context manager, so we have to enter/exit manually
+        self.stdin_swap = swap_stdin_and_argv(stdin_data=args)
+        self.stdin_swap.__enter__()
```

Comment:
Thanks @mattclay - I suspect this was copied from elsewhere, but I'll look at fixing it up today